api/src/main/java/jakarta/json/bind/annotation/JsonbPolymorphicType.java

```diff
@@ -0,0 +1,163 @@
+
+    /**
+     * Whether exact class names should be processed if no alias is specified for processed class.
+     * <br>
+     * Default value is false.
+     *
+     * @return classes without alias should be processed
+     */
+    boolean classNames() default false;
+
+    /**
+     * Whitelisted package names. This option is ignored if {@link JsonbPolymorphicType#classNames()}
+     * option is set to false.
+     * <br>
+     * Only classes contained in the selected packages will be serialized/deserialized.
+     * Classes with specified alias are not validated.
+     * <br>
+     * It is strongly recommended having whitelist set up if classes without alias should be processed.
+     * <br>
+     * When no whitelist is specified, all classes without alias are allowed.
+     *
+     * @return package whitelist
+     */
+    String[] whitelist() default "";
+
+    /**
+     * Format under which serialized polymorphic type should be handled.
+     * <br>
+     * Default polymorphic format is {@link Format#PROPERTY}.
+     */
+    enum Format {
+
+        /**
+         * Serialized object will be wrapped with another object.
+         * This wrapping object will contain only one property with the name of the
+         * polymorphic information (alias/class name) and a value assigned to this property is an actual
+         * serialized object.
+         * <br>
+         * This format ignores specified {@link JsonbPolymorphicType#keyName()}.
+         * <pre>{@code
+         * // Example
+         * @JsonbPolymorphicType(format = Format.WRAPPING_OBJECT)
+         * @JsonbSubtype(alias = "dog", type = Dog.class)
+         * interface Animal {}
+         *
+         * class Dog implements Animal {
+         *     public String isDog = true;
+         * }
+         *
+         * //Resulting json after dog instance serialization
+         * {"dog":{"isDog":true}}
+         * }</pre>
+         */
+        WRAPPING_OBJECT,
+
+        /**
+         * Serialized object will be in form of json array with two values.
+         * <br>
+         * The first in the array is polymorphic information (alias/class name) and the second
+         * array value is an actual serialized object.
+         * <br>
+         * This format ignores specified {@link JsonbPolymorphicType#keyName()}.
+         * <pre>{@code
+         * // Example
+         * @JsonbPolymorphicType(format = Format.WRAPPING_ARRAY)
+         * @JsonbSubtype(alias = "dog", type = Dog.class)
+         * interface Animal {}
+         *
+         * class Dog implements Animal {
+         *     public String isDog = true;
+         * }
+         *
+         * //Resulting json after dog instance serialization
+         * ["dog",{"isDog":true}]
+         * }</pre>
+         */
+        WRAPPING_ARRAY,
+
+        /**
+         * Serialized object will have one more property added to the resulting json.
+         * This property will contain polymorphic information (alias/class name).
```

Comment:
Do we also want to limit that this would be the first property of an object?

Reply:
Don't think so since it is not guaranteed in most JSON lib so it would make a java/JSON-B only thing almost

Reply:
@pdudits Yeah... I was thinking about that and it made sense to me to place this polymorphism information property, to be the first. But on the other hand I am not having strong opinion here, so if you think it is alright to not require this information to be the first, it is probably fine by me :-)
@rmannibucau well... Jackson does have it the same way so it would definitely not be JSON-B specific thing. So it is probably not that uncommon.

Reply:
@rmannibucau well... Jackson does have it the same way so it would definitely not be JSON-B specific thing. So it is probably not that uncommon.

Reply:
@Verdent I know it is tempting to use jackson as an example - and for this particular feature, johnzon does the same because it enables to read the type before deserializing the instance. But JSON is way wider than java and ordered serializers and we must not enforce any order IMHO because it is just what guarantees interoperability which is a way more important feature for such a solution IMHO. Also note that if we are going to make it first a custom (de)serializer does the job in 2 lines and does not require so much work so maybe the feature is not worth it at all if we start to make it very specific to our own solution (trying to share a bigger picture).

Reply:
guess ordering can respect the same than a plain inheritance as of today and consider type part of the root class.

I have a few question about the behavior in polymorphism cases:

1. What if a child of a JsonbPolymorphicType root type has another JsonbPolymorphicType? fails? ignore (think it is better to ignore).
2. Which link to `List<A>`, if we have B extends A should `List<B>` use B `JsonbPolymorphicType` A one if defined?
3. How is it linked to interfaces? We put `JsonbPolymorphicType` on interfaces? What when you get 2 interfaces implemented?

Reply:
@pdudits Sorry it took me so long to answer here. I have done it the way, that we are not requiring polymorphic information to be the first when deserializing, but it is required to be the first when object is serialized. The main reason of doing that, was for me that these polymorphic information properties are more or less metadata about the instance and therefore I didn't want to mix them with the properties of the object.

> What if key also represents regular property of the target (or parent) class?

I have had that also in mind, and I think the exception should be thrown when such situation happens? Since collision with existing property has happened. Does that make sense to you?

@rmannibucau 
> What if a child of a JsonbPolymorphicType root type has another JsonbPolymorphicType? fails? ignore (think it is better to ignore).

I think both should be in the resulting JSON. (if I am getting correctly what you mean)

> Which link to List<A>, if we have B extends A should List<B> use B JsonbPolymorphicType A one if defined?

I am sorry, but could you please elaborate a bit? Possibly with an example? I am not sure what you mean.

> How is it linked to interfaces? We put JsonbPolymorphicType on interfaces? What when you get 2 interfaces implemented?

I think interfaces should be OK to use, but we have to ensure only one JsonbPolymorphicType "chain". That means that as long as there is not more than one polymorphic information on the given level, we are fine.

This deserves a little example for better understanding I guess:
This should fail
```java
@JsonbPolymorphicType({...})
interface Machine {}

@JsonbPolymorphicType({...})
interface Vehicle {}

class Car implements Vehicle, Machine {} 
```

This should be OK
```java
@JsonbPolymorphicType({...})
interface Machine {}

@JsonbPolymorphicType({...})
interface Vehicle extends Machine {}

class Car implements Vehicle {} 
```

Reply:
> > What if a child of a JsonbPolymorphicType root type has another JsonbPolymorphicType? fails? ignore (think it is better to ignore).
> I think both should be in the resulting JSON. (if I am getting correctly what you mean)

Kind of breaks polymorphism and serialization in several case because you don't know which one to trust as a consumer (+ few other pitfalls we already listed) so not sure I agree even if the proposal looks tempting from an impl point of view.

Reply:
> List<Machine> list=List.of(new FastCar(), new Car(), new Bike())

That is actually exactly the reason why I am suggesting to include both of the polymorphic info properties. The reason is, that  you can deserialize into the `Machine` and `Vehicle` since you will be having both information present.

Machine deserialization example:
1. Machine need property `@machine` (for example)
2. Get the value of the json and map it to the aliases.
3. Vehicle class found mapped. Vehicle needs `@vehicle` property
4. Got value and mapped the alias to the Car class
5. Car class does not need any poly property required and is deserialized that way

In terms of Vehicle deserialization, it is the same, but  it starts at 3. point of the previous example, since it only needs `@vehicle` property, the other is ignored.

Doesn't that make sense? :-)

Reply:
Issue putting both is you actually loose polymorphism feature from the shape - on json side, ie you only consider java side.

So +1 to only consider very root type of the (de)serialization and not subtypes (or let it be impl dependent for now).